### PR TITLE
bridge: rewrite fsreplace1 as AsyncChannel

### PR DESCRIPTION
**Pull request by allisonkarlitskaya**
Our previous reason for avoiding tempfile.NamedTemporaryFile() was that we didn't want to chown() after the fact, but we've crossed that bridge already since 72027c1d3d5c45715030c21d6ddad18f3e06ce65, so we may as well go all the way.

Rewrite things as a single run() function using NamedTemporaryFile as a context manager.  This is a fairly substantial simplification.  We also make the channel properly asynchronous by dispatching the actual write to an executor.  We also add fdatasync(), which we were missing before.

### CODE REVIEW

File: src/cockpit/channels/filesystem.py [hunk 1]
```diff
@@ -50,6 +52,12 @@
 logger = logging.getLogger(__name__)
 
 
+@functools.lru_cache()
+def my_umask() -> int:
+    match = re.search(r'^Umask:\s*0([0-7]*)$', Path('/proc/self/status').read_text(), re.M)
```
Comment (martinpitt) [inline src/cockpit/channels/filesystem.py:57]: Ugh, this is horrible.. I know that [os.umask](https://docs.python.org/3/library/os.html#os.umask) is weird too (due to `umask(2)`). Calling [getumask()](https://www.man7.org/linux/man-pages/man3/getumask.3.html) via ctypes is equally involved, though. And NamedTemporaryFile() doesn't have a mode where it would just open the file normaly. At times, Linux APIs are really shitty..

File: src/cockpit/channels/filesystem.py [hunk 2]
```diff
@@ -150,110 +158,75 @@ def do_yield_data(self, options: JsonObject) -> GeneratorChannel.DataGenerator:
...
-            self._tempfile.close()
-            self._tempfile = None
+                # otherwise, spool data into a temporary file until EOF then rename into place...
+                with tempfile.NamedTemporaryFile(dir=dirname, prefix=f'.{basename}-', delete=False) as tmp:
+                    loop = asyncio.get_running_loop()
+                    while data is not None:
+                        await loop.run_in_executor(None, tmp.write, data)
+                        data = await self.read()
+
+                    os.fdatasync(tmp.fileno())
```
Comment (martinpitt) [inline src/cockpit/channels/filesystem.py:188]: Why is that necessary? The commit message doesn't give the "why".

File: src/cockpit/channels/filesystem.py [hunk 3]
```diff
@@ -150,110 +158,75 @@ def do_yield_data(self, options: JsonObject) -> GeneratorChannel.DataGenerator:
...
+                    else:
+                        # the file must exist with the given tag
+                        buf = os.stat(path)
+                        if tag != tag_from_stat(buf):
+                            raise ChannelError('change-conflict')
+                        # chown/chmod from the existing file permissions
+                        os.fchmod(tmp.fileno(), stat.S_IMODE(buf.st_mode))
+                        os.fchown(tmp.fileno(), buf.st_uid, buf.st_gid)
+                        os.rename(tmp.name, path)
+                        tmp = None  # don't need to delete
```
Comment (martinpitt) [inline src/cockpit/channels/filesystem.py:211]: Wow, is that safe with a `with` statement?  This looks like it could seriously crash the contex's `__exit__()`. It apparently doesn't right now, but this is too hackish for my taste.

I just tested
```
>>> import tempfile, os
>>> with tempfile.NamedTemporaryFile() as tmp:
...     os.unlink(tmp.name)
```
and that works fine. There is no situation where we would *ever* want to keep the temporary file around, it should be removed in all cases. So it seems that can be cleaned up?

File: src/cockpit/channels/filesystem.py [hunk 4]
```diff
@@ -150,110 +158,75 @@ def do_yield_data(self, options: JsonObject) -> GeneratorChannel.DataGenerator:
...
-                # not ideal, but the closest code we have
-                raise ChannelError('access-denied', message=str(exc)) from exc
-            except OSError as exc:
-                self.unlink_temppath()
-                raise ChannelError('internal-error', message=str(exc)) from exc
-
-            self._tempfile.close()
-            self._tempfile = None
+                # otherwise, spool data into a temporary file until EOF then rename into place...
+                with tempfile.NamedTemporaryFile(dir=dirname, prefix=f'.{basename}-', delete=False) as tmp:
```
Comment (martinpitt) [inline src/cockpit/channels/filesystem.py:182]: The context `tmp` here looks like it shadows the `tmp` from above which his already in scope. I experimented with this and it seems it actually merely overwrites it, and `tmp` is still available after the `with`. But still, this is hard to follow and unclean.

File: src/cockpit/channels/filesystem.py [hunk 5]
```diff
@@ -150,110 +158,75 @@ def do_yield_data(self, options: JsonObject) -> GeneratorChannel.DataGenerator:
...
+        except PermissionError as exc:
+            raise ChannelError('access-denied') from exc
+        except IsADirectoryError as exc:
+            # not ideal, but the closest code we have
+            raise ChannelError('access-denied', message=str(exc)) from exc
+        except OSError as exc:
+            raise ChannelError('internal-error', message=str(exc)) from exc
+        finally:
+            if tmp is not None:
+                os.unlink(tmp.name)
```
Comment (martinpitt) [inline src/cockpit/channels/filesystem.py:227]: See above -- doing that should be the whole point of using a `with` context? I seriously doubt my Python knowledge right now..

**Review by martinpitt — COMMENTED** [review]
Thanks! I suppose the conversion to asyncchannel is necessary to support large file uploads in c-files? That part is nice.

I'm less sure about the NamedTemporaryFile, though. The "compute temp path" code is/was gross, but the umask and `tmp` scoping hacks are as well. I have some questions about it, but if the `with` context seriously doesn't clean up when exceptions happen, it's not worth having IMHO.

File: src/cockpit/channels/filesystem.py (hunk 2 above)
Comment (allisonkarlitskaya) [reply src/cockpit/channels/filesystem.py]: If we don't call fdatasync() then we're not guaranteed to have the new file's data on disk before the rename takes effect, and we could end up with a zero-length file.

This is how `g_file_set_contents()` does it, for example.  See this comment: https://gitlab.gnome.org/GNOME/glib/-/blob/704d0bb297cff6fc53beca48ccc2cfd061f0bf5e/glib/gfileutils.c#L1094

File: src/cockpit/channels/filesystem.py (hunk 3 above)
Comment (allisonkarlitskaya) [reply src/cockpit/channels/filesystem.py:211]: Does that also work in Python 3.6?  At some point this gave `FileNotFoundError`.  They recently improved the situation with a new `delete_on_close` kwarg in Python 3.12, so I wouldn't be surprised if there were other cleanups on this point.

We could do the delete and suppress the message, but there's a theoretical (albeit effectively 0) possibility that we're deleting a file that we didn't create in that case:

 - we create the tempfile, write write, fsync
 - rename it into place over the real filename
 - at this exact moment another process comes along and creates a new temp file and just happens to get exactly the same name as the one we used
 - we `os.unlink()` what we think is our filename, but is actually the one created by the other process

So effectively, I'm using `tmp = None` here to mean "don't delete this".

As for the context manager vs. `None` thing, this isn't actually a problem.  There are two things involved in a context manager:

 - the context manager itself (which is captured by the `with` block to have `__exit__()` called later)
 - the return value of the `__enter__()` function of the context manager — this is the thing that gets optionally captured by the optional `as` clause

We're nulling out the second copy, but not the first.  In this case, I *think* it's the same object, but to be honest I'm not sure.  In either case, it doesn't matter — `with` holds its own copy and will `.close()` it (but not delete it).

I had a different version of this code that kept a separate variable which was the name of the file that we should delete when exiting the function.  I could try that again if you feel it is better.  I ended up liking this version more.

File: src/cockpit/channels/filesystem.py (hunk 4 above)
Comment (allisonkarlitskaya) [reply src/cockpit/channels/filesystem.py:182]: Ya.  I buy that argument.  It was also a bit difficult to convince `mypy` to like this.

File: src/cockpit/channels/filesystem.py (hunk 5 above)
Comment (allisonkarlitskaya) [reply src/cockpit/channels/filesystem.py:227]: We create the `NamedTempFile()` with `delete=True`, which means that the contextmanager will close the tempfile, but not delete it.  This would all be so much better if we could use `delete_on_close=True`, but it's only in 3.12 :(

File: src/cockpit/channels/filesystem.py (hunk 2 above)
Comment (martinpitt) [reply src/cockpit/channels/filesystem.py]: Ah, so this isn't about rename() not working on "only in write cache" data (that would surprise me, as it's entirely an inode-level operation), but just generally making sure about data integrity, i.e. strong promises about that API ("if it succeeds, it's on disk"). This isn't documented. I don't mind adding it, it just wasn't clear why it's necessary for the renaming to work (... as it isn't). Thanks! (I'd prefer a second commit with an explanation, but not insisting)

File: src/cockpit/channels/filesystem.py (hunk 5 above)
Comment (martinpitt) [reply src/cockpit/channels/filesystem.py:227]: IMHO `delete_on_close=True` is the whole reason why we want to use NamedTemporaryFile. I.e. *always* remove the temp file, and just rename it to the final name in the happy path. I'd consider a rewrite using `=False` as making it worse..

File: src/cockpit/channels/filesystem.py (hunk 3 above)
Comment (martinpitt) [reply src/cockpit/channels/filesystem.py:211]: Thanks for the explanation. Makes sense to keep an internal copy, so that the code can mess around with the variable. 

I tested it on py3.6 (rhel 8.1.0) and that indeed fails with a `FileNotFoundError`.

All that umask/weird scoping here feels like that NamedTemporaryFile isn't really the API that we want? Replacing one hack (tmpname calculation) with two others (smaller, but harder to understand) doesn't really feel like an improvement to me.

Just to gauge: I don't like it, but not a veto. I just wanted to understand how this works. If you like it, then go ahead, but I wouldn't mind keeping the current approach for this (the async conversion is really nice, though, but that's independent).

**Comment by martinpitt** [comment]
Looking at the [dump unit test failure](https://github.com/cockpit-project/cockpit/actions/runs/8614448618/job/23607901576?pr=20278#step:2:779):
```
 log: ok 6 - config_update: raw text for changed config is correct
> log: not ok 7 - global failure: Error: Uncaught TypeError: Cannot read properties of null (reading '_internal'), test: config_update, module: kdump, source: Error: Uncaught TypeError: Cannot read properties of null (reading '_internal')
#     at window$1.onerror (http://localhost:44169/qunit/kdump/test-config-client.js:5828:39)
#     at window.onerror (http://localhost:44169/qunit/kdump/test-config-client.js:9533:16)
```

It reproduces fine with `./test-server` and http://localhost:8765/qunit/kdump/test-config-client.html , or with `pytest -k kdump/test-config-client.html --timeout=5` (the default 2 min qunit timeout is annoying otherwise).

The crash happens because the `configChanged` callback is called twice now. First with the expected

```
{"_internal":"Object","targets":"Object","compression":"Object","core_collector":"makedumpfile -l --message-level 7 -d 31"}
```
but the second time *after* the whole test is done and calling `config.close()` -- then it gets called with `null`. That goes away when I remove the cleanup hook:
```js
    hooks.after(() => cockpit.spawn(["rm", "-f", filename]));
```
i.e. the problem is that the watch somehow sticks around even after the watch channel got closed with `config.close()`. Note that this cleanup hook really should use `cockpit.file.replace(null)`, but in this case it actually works in our favor as it rules out the fsreplace1 channel as the culprit.

This primarily concerns fswatch, not fsreplace, but it can't be a concidence that making this channel async interferes with fswatch?

I'll look at this some more and then update the comment -- but this is subtle.

File: src/cockpit/channels/filesystem.py (hunk 1 above)
Comment (allisonkarlitskaya) [reply src/cockpit/channels/filesystem.py:57]: The ctypes call isn't bad.  I knew there was nothing at the kernel level for this, but a threadsafe-via-lock equivalent in the C library isn't bad!  Is that on RHEL8?

Comment (allisonkarlitskaya) [reply src/cockpit/channels/filesystem.py:57]:
```
⬢[lis@toolbox ~]$ podman run -it ubi8
[root@c4818a63b310 /]# dnf install python3
...
[root@c4818a63b310 /]# python3
Python 3.6.8 (default, Jan  5 2024, 08:58:17) 
[GCC 8.5.0 20210514 (Red Hat 8.5.0-20)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import ctypes
>>> c = ctypes.CDLL(None)
>>> c.getumask()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib64/python3.6/ctypes/__init__.py", line 356, in __getattr__
    func = self.__getitem__(name)
  File "/usr/lib64/python3.6/ctypes/__init__.py", line 361, in __getitem__
    func = self._FuncPtr((name_or_ordinal, self))
AttributeError: python3: undefined symbol: getumask
```

bzzt

File: src/cockpit/channels/filesystem.py (hunk 2 above)
Comment (allisonkarlitskaya) [reply src/cockpit/channels/filesystem.py]: It's not yet clear to me that we accept this proposal PR in principle yet, so it would indeed be useful to have this if we stay with the old approach.

I'd feel pretty bad about calling that outside of a thread, though — it can block for seconds... and calling it in a thread in a way that works in callback-based code would start to get a bit hairy...

File: src/cockpit/channels/filesystem.py (hunk 3 above)
Comment (allisonkarlitskaya) [reply src/cockpit/channels/filesystem.py:211]: So there's another consideration here which I didn't explicitly state that made me lean more towards `NamedTemporaryFile`.  I *really* dislike the chown/chmod tricks (as you know), and I particularly dislike the idea that we might be in a situation where either:
 - we chown/chmod late, in which case we might temporarily reveal sensitive data which ought not to be revealed (which is a security problem)
 - we chown/chmod early, in which case we might have some file with special bits like setuid in a half-written state (which is a security problem)

So if we're going to `chown/chmod`, then I specifically like that `NamedTemporaryFile` creates a file as the current user, with a maximally locked down set of permissions, and we only change the file mode/owner after it's fully written.  It actually legit seems like the best way.

But ya — the inability to deal with umask nicely from Python is a long-documented and awful problem.  There's a similar problem where it's very difficult to create executable files with the proper permissions because `open()` doesn't support the executable bit and you can't `chmod()` after the fact because of the frustration of determining the value of the umask.  The solution in that case is the same, and it sucks — lose all of the benefits of the standard library and write low-level code via `os.open()`.

File: src/cockpit/channels/filesystem.py (hunk 5 above)
Comment (allisonkarlitskaya) [reply src/cockpit/channels/filesystem.py:227]: From where I'm sitting: context managers are nice, but so are `finally:` clauses.  They're essentially interchangeable.

`NamedTemporaryFile`, in the best case, does ~4 things:
 - creates a unique name
 - gives you a file-like object for interacting with it
 - closes it when you're done
 - deletes it when you're done

The last one is not working for us, so we have to go manual on that one.... but I'm happy to use the other 3...

File: src/cockpit/channels/filesystem.py (hunk 1 above)
Comment (martinpitt) [reply src/cockpit/channels/filesystem.py:57]: It was more of a lament than a request to change, but thanks for trying :grin:

Comment (martinpitt) [reply src/cockpit/channels/filesystem.py:57]: It doesn't work on Fedora 39 either, btw

File: src/cockpit/channels/filesystem.py (hunk 3 above)
Comment (martinpitt) [reply src/cockpit/channels/filesystem.py:211]: Note: we can do the "start with minimal permissions" fairly easily in the current code as well if we change it to
```
fd = os.open(..., 0o666 if stat is None else 0600)
```
and moving the chown to closing.

File: src/cockpit/channels/filesystem.py (hunk 5 above)
Comment (martinpitt) [reply src/cockpit/channels/filesystem.py:227]: Thanks for evaluating all the options. So the summary is "they all suck, Linux doesn't have a good API for this, let's just pick the least bad one"  :sweat_smile: 

So, I'm ok with this, we just need to figure out the fsreplace/fswatch race below.

**Comment by martinpitt** [comment]
A-ha! With more debugging around cockpit.js I see that this happens:
```
cockpit.protocol-DEBUG: channel control received {'problem': 'cancelled', 'command': 'close', 'channel': '1:7'}
> log: XXXX try_read problem {"problem":"cancelled","command":"close","channel":"1:7"}
```

So the issue is that `.close()`'ing the channel now gets a "cancelled" problem, which `try_read()` interprets as a read error and calls the watch callback with `null`:
```js
                    if (message.problem) {
                        const error = new BasicError(message.problem, message.message);
                        console.log("XXXX try_read problem", JSON.stringify(message));
                        fire_watch_callbacks(null, null, error);
                        dfd.reject(error);
                        return;
                    }
```

From the perspective of `cockpit.file.read()` this is actually correct -- it's just a bit awkward in the context of a watch channel with implied reading. At first sight this sounds related to the channel.py change here, but reverting just that gives the same result. WIth reverting the whole patch, of course the channel also still receives the "cancelled" problem, but it doesn't hit `try_read()`. I don't see an obvious reason why, though, except sheer luck with timing.

It certainly *would* be nice if we could guarantee that calling `.close()` on a watch channel would never invoke the callbacks any more. But with everything being async and deferred, that may be hard to do. Failing that, we could apply this robustification:

```diff
--- pkg/kdump/test-config-client.js
+++ pkg/kdump/test-config-client.js
@@ -83,6 +83,7 @@ QUnit.module("kdump", hooks => {
                         config.write(config.settings)
                                 .then(() => {
                                     // Close watch channel
+                                    config.removeEventListener('kdumpConfigChanged', configChanged);
                                     config.close();
                                     dataWasChanged.then(done);
```

With that, the test works reliably.

**Comment by allisonkarlitskaya** [comment]
So this unfortunately landed already, as part of #20315.  I ignored the red testing-farm checks on CentOS 9 there because I figured that it wasn't possible that they were relevant to the work there, but this change is quite a lot more invasive, and the same checks are red here.  Let's re-run those to find out if this is what caused that....

**Comment by allisonkarlitskaya** [comment]
... yup, looks like testing farm was just flaking.

So then this is taken care of in https://github.com/cockpit-project/cockpit/pull/20315.
